patches/tModLoader/Terraria/GameContent/TeleportPylonsSystem.cs.patch

```diff
@@ -1,11 +1,209 @@
++			_moddedPylons.Clear();
+ 			foreach (TileEntity value in TileEntity.ByPosition.Values) {
+-				TETeleportationPylon tETeleportationPylon = value as TETeleportationPylon;
+-				if (tETeleportationPylon != null && tETeleportationPylon.TryGetPylonType(out TeleportPylonType pylonType)) {
+-					TeleportPylonInfo teleportPylonInfo = default(TeleportPylonInfo);
+-					teleportPylonInfo.PositionInTiles = tETeleportationPylon.Position;
++				TeleportPylonType pylonType = TeleportPylonType.Modded;
++				if (value.GetType().IsDefined(typeof(IsPylonEntityAttribute), true) || (value is TETeleportationPylon tETeleportationPylon && tETeleportationPylon.TryGetPylonType(out pylonType)))  {
++					TeleportPylonInfo teleportPylonInfo = new TeleportPylonInfo();
++					teleportPylonInfo.PositionInTiles = value.Position;
+ 					teleportPylonInfo.TypeOfPylon = pylonType;
++					if (pylonType == TeleportPylonType.Modded) {
++						Tile tile = Framing.GetTileSafely(value.Position);
++						ModPylon pylon = TileLoader.GetTile(tile.type) as ModPylon;
++						teleportPylonInfo.ModName = pylon!.Mod.Name;
++						teleportPylonInfo.ModPylonName = pylon!.Name;
++					}
+ 					TeleportPylonInfo item = teleportPylonInfo;
+-					_pylons.Add(item);
++					AddNewPylon(item);
+ 				}
+ 			}
+ 
+@@ -67,12 +_,14 @@
+ 		}
+ 
+ 		public void AddForClient(TeleportPylonInfo info) {
+-			if (!_pylons.Contains(info))
+-				_pylons.Add(info);
++			if (!_pylons.Contains(info)) 
++				AddNewPylon(info);
+ 		}
+ 
+ 		public void RemoveForClient(TeleportPylonInfo info) {
+-			_pylons.RemoveAll((TeleportPylonInfo x) => x.Equals(info));
++			_pylons.RemoveAll(x => x.Equals(info));
++			_vanillaPylons.RemoveAll(x => x.Equals(info));
++			_moddedPylons.RemoveAll(x => x.Equals(info));
+ 		}
+ 
+ 		public void HandleTeleportRequest(TeleportPylonInfo info, int playerIndex) {
+@@ -93,7 +_,14 @@
+ 			}
+ 
+ 			if (flag) {
+-				flag &= !NPC.AnyDanger(quickBossNPCCheck: false, ignorePillars: true);
++				bool forceSucceed = false;
++				if (PylonLoader.ValidTeleportCheck_PreAnyDanger(info) is { } value) {
++					forceSucceed = value;
++					flag &= value;
++				}
++				flag &= (info.TypeOfPylon == TeleportPylonType.Modded
++					? ModContent.TryFind(info.ModName, info.ModPylonName, out ModPylon pylon) && pylon.ValidTeleportCheck_AnyDanger(info)
++					: !NPC.AnyDanger(quickBossNPCCheck: false, ignorePillars: true)) || forceSucceed;
+ 				if (!flag)
+ 					key = "Net.CannotTeleportToPylonBecauseThereIsDanger";
+ 			}
+@@ -120,9 +_,11 @@
+ 				if (!flag)
+ 					key = "Net.CannotTeleportToPylonBecauseNotMeetingBiomeRequirements";
+ 			}
++			if (info.TypeOfPylon == TeleportPylonType.Modded && ModContent.TryFind(info.ModName, info.ModPylonName, out ModPylon destinationPylon))
++				destinationPylon.ValidTeleportCheck_DestinationPostCheck(info, ref flag, ref key);
+ 
+-			if (flag) {
+-				bool flag2 = false;
++			TeleportPylonInfo nearbyInfo = new TeleportPylonInfo();
++			bool flag2 = false;
```

Comment:
Just keep a placeholder brace for indentation and patch maintenance.
Also do `if (flag || true) // tML: check for destination pylon even if no source pylon`